app/models/transformation_mapping/vm_migration_validator.rb

```diff
@@ -0,0 +1,146 @@
+    vm_objects = Vm.where(:name => vm_names, :ems_cluster => mapped_clusters).to_a
+    MiqPreloader.preload(vm_objects, %i(lans storages host ext_management_system))
+
+    @vm_list.each do |row|
+      vm_name = row['name']
+
+      if vm_name.blank?
+        invalid_list << VmMigrateStruct.new('', nil, VM_INVALID, VM_EMPTY_NAME)
+        next
+      end
+
+      vms = vm_objects.select do |vm|
+        found = vm.name == vm_name
+        found = vm.uid_ems == row['uid_ems'] if found && row['uid_ems'].present?
+        found = vm.host == row['host'] if found && row['host'].present?
+        found = vm.ext_management_system == row['provider'] if found && row['provider'].present?
+        found
+      end
+
+      if vms.size.zero?
+        invalid_list << VmMigrateStruct.new(vm_name, nil, VM_INVALID, VM_NOT_EXIST)
+      elsif vms.size == 1
+        vm = vms.first
+        reason = validate_vm(vm, false)
+        if reason == VM_VALID
+          valid_list << VmMigrateStruct.new(vm.name, vm, VM_VALID, reason)
+        else
+          invalid_list << VmMigrateStruct.new(vm.name, vm, VM_INVALID, reason)
+        end
+      else
+        vms.each { |v| conflict_list << VmMigrateStruct.new(v.name, v, VM_CONFLICT, VM_CONFLICT) }
+      end
+    end
+
+    {
+      "valid"      => valid_list,
+      "invalid"    => invalid_list,
+      "conflicted" => conflict_list
+    }
+  end
+
+  def validate_vm(vm, quick = true)
+    # a valid vm must find all resources in the mapping and has never been migrated
+    invalid_list = []
+
+    invalid_storages = vm.datastores - mapped_storages
+    if invalid_storages.present?
+      invalid_list << "storages: %{list}" % {:list => invalid_storages.collect(&:name).join(", ")}
+      return no_mapping_msg(invalid_list) if quick
+    end
+
+    invalid_lans = vm.lans - mapped_lans
+    if invalid_lans.present?
+      invalid_list << "lans: %{list}" % {:list => invalid_lans.collect(&:name).join(', ')}
+      return no_mapping_msg(invalid_list) if quick
+    end
+
+    return no_mapping_msg(invalid_list) if invalid_list.present?
+    vm_migration_status(vm)
+  end
+
+  def vm_migration_status(vm)
+    vm_as_resources = ServiceResource.joins(:service_template).where(:resource => vm, :service_templates => {:type => 'ServiceTemplateTransformationPlan'})
+
+    # VM has not been migrated before
+    return VM_VALID if vm_as_resources.empty?
+
+    return VM_MIGRATED unless vm_as_resources.where(:status => ServiceResource::STATUS_COMPLETED).empty?
+
+    # VM failed in previous migration
+    vm_as_resources.all? { |rsc| rsc.status == ServiceResource::STATUS_FAILED } ? VM_VALID : VM_IN_OTHER_PLAN
+  end
+
+  def no_mapping_msg(list)
+    "Mapping source not found - %{list}" % {:list => list.join('. ')}
+  end
+
+  def mapped_clusters
+    @mapped_clusters ||= EmsCluster.where(:id => @mapping.transformation_mapping_items.where(:source_type => 'EmsCluster').pluck(:source_id))
```

Comment:
please change `pluck` to select`

it will do 1 query instead of 2 here

Reply:
@kbrock Isn't the `select` slower overall since it instantiates ruby objects for each result rather than just the `source_id`s?  

Also, it looks like they're both 2 queries...

```ruby
$ rails c
Loading development environment (Rails 5.0.7)
irb(main):001:0> PxeImageType.first.pxe_images.select(:name)
PostgreSQLAdapter#log_after_checkout, connection_pool: size: 5, connections: 1, in use: 1, waiting_in_queue: 0
  PxeImageType Load (0.7ms)  SELECT  "pxe_image_types".* FROM "pxe_image_types" ORDER BY "pxe_image_types"."id" ASC LIMIT $1  [["LIMIT", 1]]
  PxeImageType Inst Including Associations (9.7ms - 1rows)
  PxeImage Load (0.4ms)  SELECT "pxe_images"."name" FROM "pxe_images" WHERE "pxe_images"."pxe_image_type_id" = $1  [["pxe_image_type_id", 1]]
  PxeImage Inst Including Associations (8.8ms - 1rows)
(Object doesn't support #inspect)
=> 
irb(main):002:0> exit

$ rails c
Loading development environment (Rails 5.0.7)
irb(main):001:0> PxeImageType.first.pxe_images.pluck(:name)
PostgreSQLAdapter#log_after_checkout, connection_pool: size: 5, connections: 1, in use: 1, waiting_in_queue: 0
  PxeImageType Load (0.5ms)  SELECT  "pxe_image_types".* FROM "pxe_image_types" ORDER BY "pxe_image_types"."id" ASC LIMIT $1  [["LIMIT", 1]]
  PxeImageType Inst Including Associations (6.0ms - 1rows)
   (0.3ms)  SELECT "pxe_images"."name" FROM "pxe_images" WHERE "pxe_images"."pxe_image_type_id" = $1  [["pxe_image_type_id", 1]]
=> ["abc"]
irb(main):002:0> exit
```

Reply:
```ruby
EmsCluster.where( # 1
  :id => @mapping.transformation_mapping_items.where(:source_type => 'EmsCluster').pluck(:source_id) # 2
)
```

If `#2` uses a `pluck`, a query is run for `#2` and a query for `#1`. -- 2 queries total.
If `#2` uses a `select`, the `#1` query will have `#2` as a subquery. -- 1 query total.

Reply:
Seems the query here has been changed that #1 query does not exist any more. So no matter using ```pluck``` or ```select```, it is always one query. I would prefer to use ```pluck``` in this case.

@bdunne @kbrock Anything else to update? Thanks.

Reply:
I still request this change (there are 20 vm records in my example, 1 ems)

Using `pluck` forces all `id` values to come back from postgres, and then send a separate query to the server with all the values:

```ruby
ExtManagementSystem.where(:id => VmOrTemplate.where("name like 'VC0DC0_C0_RP1%'").pluck(:ems_id)).to_a
   (1.1ms)  SELECT "vms"."ems_id" FROM "vms" WHERE (name like 'VC0DC0_C0_RP1%')
  ExtManagementSystem Load (0.5ms)  SELECT "ext_management_systems".* FROM "ext_management_systems" WHERE "ext_management_systems"."id" IN (1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1)
  ExtManagementSystem Inst Including Associations (0.3ms - 1rows)
```

Using `select` allows the block to be used in an inner select - fewer round trips, less logic to create the sql, many fewer local objects:

```ruby
ExtManagementSystem.where(:id => VmOrTemplate.where("name like 'VC0DC0_C0_RP1%'").select(:ems_id)).to_a
  ExtManagementSystem Load (1.1ms)  SELECT "ext_management_systems".* FROM "ext_management_systems" WHERE "ext_management_systems"."id" IN (SELECT "vms"."ems_id" FROM "vms" WHERE (name like 'VC0DC0_C0_RP1%'))
  ExtManagementSystem Inst Including Associations (0.4ms - 1rows)
```

Reply:
With ```pluck```:
```
EmsCluster.where(:id => mapping.transformation_mapping_items.where(:source_type => 'EmsCluster').pluck(:source_id))
   (0.3ms)  SELECT "transformation_mapping_items"."source_id" FROM "transformation_mapping_items" WHERE "transformation_mapping_items"."transformation_mapping_id" = $1 AND "transformation_mapping_items"."source_type" = $2  [["transformation_mapping_id", 1], ["source_type", "EmsCluster"]]
  EmsCluster Load (0.5ms)  SELECT "ems_clusters".* FROM "ems_clusters" WHERE "ems_clusters"."id" IN (24, 12)
  EmsCluster Inst Including Associations (0.3ms - 2rows)
```
With ```select```:
```
EmsCluster.where(:id => mapping.transformation_mapping_items.where(:source_type => 'EmsCluster').select(:source_id))
  EmsCluster Load (0.9ms)  SELECT "ems_clusters".* FROM "ems_clusters" WHERE "ems_clusters"."id" IN (SELECT "transformation_mapping_items"."source_id" FROM "transformation_mapping_items" WHERE "transformation_mapping_items"."transformation_mapping_id" = $1 AND "transformation_mapping_items"."source_type" = $2)  [["transformation_mapping_id", 1], ["source_type", "EmsCluster"]]
  EmsCluster Inst Including Associations (89.5ms - 2rows)
```

Reply:
thanks Lucy
I do agree with @bdunne that the majority of cases `pluck` is better than `select`.
This case is just a little different because we weren't doing anything with the ids.